sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -116,13 +116,26 @@ object FileFormatWriter extends Logging {
     val partitionSet = AttributeSet(partitionColumns)
     val dataColumns = allColumns.filterNot(partitionSet.contains)
 
-    val bucketIdExpression = bucketSpec.map { spec =>
+    val bucketPartitioning = bucketSpec.map { spec =>
       val bucketColumns = spec.bucketColumnNames.map(c => dataColumns.find(_.name == c).get)
+      HashPartitioning(bucketColumns, spec.numBuckets)
+    }
+
+    val bucketIdExpression = bucketPartitioning.map { partitioning =>
       // Use `HashPartitioning.partitionIdExpression` as our bucket id expression, so that we can
       // guarantee the data distribution is same between shuffle and bucketed data source, which
       // enables us to only shuffle one side when join a bucketed table and a normal one.
-      HashPartitioning(bucketColumns, spec.numBuckets).partitionIdExpression
+      partitioning.partitionIdExpression
     }
+
+    // If the plan's outputPartitioning is the same as the the bucket spec, then each row will have
+    // a constant bucket id. We possibly can avoid the sort altogether.
+    val bucketSortExpression = plan.outputPartitioning match {
+      case output: HashPartitioning =>
+        bucketPartitioning.filterNot(_.semanticEquals(output)).map(_.partitionIdExpression)
```

Comment:
If bucket expressions are the same as the partition expressions and the number of buckets are the same as the number of partitions, this may work. In this case, the bucket id is actually the partition id.

Actually, I'm doubt that is this useful? You will have one file per partition because all rows in a partition are in the same bucket. So basically you don't have any advantage of bucket, right?